core/src/main/c/share/jit/avx2.h

```diff
@@ -205,10 +208,49 @@ namespace questdb::avx2 {
         return {length_data, data_type_t::i64, data_kind_t::kMemory};
     }
 
+    // Reads length part of the varchar header for aux vector.
+    // This part is stored in the lowest bytes of the header
+    // (see VarcharTypeDriver to understand the format).
+    //
+    // Note: unlike read_mem_varsize this method doesn't return the length,
+    //       so it can only be used in NULL checks.
+    jit_value_t read_mem_varchar_header(Compiler &c,
+                                        int32_t column_idx,
+                                        const Gp &varsize_aux_ptr,
+                                        const Gp &input_index) {
+        Gp varsize_aux_address = c.newInt64("varsize_aux_address");
+        c.mov(varsize_aux_address, ptr(varsize_aux_ptr, 8 * column_idx, 8));
+
+        Gp header_offset_0 = c.newInt64("header_offset_0");
+        Gp header_offset_2 = c.newInt64("header_offset_2");
+
+        c.mov(header_offset_0, input_index);
+        auto header_shift = type_shift(data_type_t::i128);
+        c.sal(header_offset_0, header_shift);
+        c.mov(header_offset_2, header_offset_0);
+        c.add(header_offset_2, 32);
+
+        Ymm headers_0_1 = c.newYmm("headers_0_1");
+        Ymm headers_2_3 = c.newYmm("headers_2_3");
+
+        // Load 4 headers into two YMMs.
+        c.vmovdqu(headers_0_1, ymmword_ptr(varsize_aux_address, header_offset_0, 0));
+        c.vmovdqu(headers_2_3, ymmword_ptr(varsize_aux_address, header_offset_2, 0));
+
+        // Permute the first i64 of each header to be in the first YMM lane
+        // and combine them into single YMM.
+        c.vpermq(headers_0_1, headers_0_1, 0b00001000);
+        c.vpermq(headers_2_3, headers_2_3, 0b00001000);
+        c.vinserti128(headers_0_1, headers_0_1, headers_2_3.xmm(), 1);
```

Comment:
My understanding is that all three of these ops are cross-lane: 

1. Gather headers 0 and 1 to XMM of headers_0_1
2. Gather headers 2 and 3 to XMM of headers_2_3
3. Insert the XMM lane of headers_2_3 to the upper lane of headers_2_3

Can we perhaps cheaply improve this by removing one cross-lane operation? 

Like this: gather headers 2 and 3 directly to the upper lane of headers_2_3, and then copy the XMM of headers_0_1 to headers_2_3.

Reply:
I considered that, but on the other hand, operations 1 and 2 are independent, so they can run in parallel. This means that the overall latency should be the same as with the current operations. Still, let me try that and measure if there is any difference.

Reply:
Sorry, I misread your suggestion. You propose to get rid of cross-lane copying in op 3. Going to try that.

Reply:
Gave it a try in fcc6b5b397a3e63d00ca683b5c6a3b604cd34ee5 PTAL

The improvement is small, if any (since it could be jitter), but I'll keep this version.

Before:
```
Benchmark                                                  (a_nullFreq)  (b_useComplexFilter)  (c_jitMode)  Mode  Cnt    Score    Error  Units
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             0                 false         SIMD  avgt    5   13.648 ±  0.389  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             0                  true         SIMD  avgt    5   17.632 ±  0.436  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             1                 false         SIMD  avgt    5   35.267 ±  1.539  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             1                  true         SIMD  avgt    5   45.388 ±  0.823  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             3                 false         SIMD  avgt    5   31.777 ±  0.533  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             3                  true         SIMD  avgt    5   42.229 ±  0.951  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull           100                 false         SIMD  avgt    5   14.974 ±  0.351  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull           100                  true         SIMD  avgt    5   19.836 ±  0.367  ms/op
```

After:
```
Benchmark                                                  (a_nullFreq)  (b_useComplexFilter)  (c_jitMode)  Mode  Cnt   Score   Error  Units
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             0                 false         SIMD  avgt    5  12.907 ± 1.750  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             0                  true         SIMD  avgt    5  17.344 ± 3.641  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             1                 false         SIMD  avgt    5  33.954 ± 1.154  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             1                  true         SIMD  avgt    5  44.665 ± 4.612  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             3                 false         SIMD  avgt    5  30.702 ± 1.353  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull             3                  true         SIMD  avgt    5  40.807 ± 1.125  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull           100                 false         SIMD  avgt    5  14.635 ± 1.919  ms/op
StringNullCheckBenchmark.selectCountWhereVarcharColIsNull           100                  true         SIMD  avgt    5  19.344 ± 2.613  ms/op
```